File: singly_linked_list.go
package main

import "fmt"

type Node struct {
	data int
	next *Node
}

var head *Node

func Insert(x int) {
	node := new(Node)
	node.data = x
	node.next = head
	head = node
}

func Reverse(node *Node) {
	if node.next == nil {
		head = node
		return
	}

	Reverse(node.next)

	node.next.next = node
	node.next = nil
}

func Print() {
	var node *Node = head
	fmt.Println("List is: ")
	for node != nil {
		fmt.Println(node.data)
		node = node.next
	}
	fmt.Println("\n")
}

func main() {
	head = nil
	var n, x int

	fmt.Println("How many numbers?")
	fmt.Scan(&n)

	i := 0
	for i < n {
		fmt.Println("Enter the number")
		fmt.Scan(&x)
		Insert(x)
		Print()
		i++
	}

	Reverse(head)
	Print()
}

File: singly_linked_list.c
#include <stdlib.h>
#include <stdio.h>

struct Node {
  int data;
  struct Node* next;
};

struct Node* head;

void Insert(int x) {
  struct Node* newNode = malloc(sizeof(struct Node));
  newNode->data = x;
  newNode->next = head;
  head = newNode;
}

void Reverse(struct Node* node) {
  if (node->next == NULL) {
    head = node;
    return;
  }

  Reverse(node->next);

  struct Node* nodeNext = node->next;
  node->next->next = node;
  node->next = NULL;
}

void Print() {
  struct Node* firstNode = head;
  printf("List is: ");
  while (firstNode != NULL) {
    printf(" %d ", firstNode->data);
    firstNode = firstNode->next;
  }
  printf("\n");
}

int main() {
  head = NULL;
  printf("How many numbers?\n");
  int n, i, x;
  scanf("%d", &n);
  for (i = 0; i < n; i++) {
    printf("Enter the number\n");
    scanf("%d", &x);
    Insert(x);
    Print();
  }

  Reverse(head);
  Print();
}
